include/dice/template-library/polymorphic_allocator.hpp

```diff
@@ -314,33 +318,91 @@ namespace dice::template_library {
 
 #if __has_include(<boost/interprocess/offset_ptr.hpp>)
 	/**
-	 * @brief Basically std::allocator but returning boost::interprocess::offset_ptr
+	 * @brief Wraps and `std::allocator`-like type, but returns boost::interprocess::offset_ptr instead of raw pointers
 	 * @tparam T type to allocate
 	 */
-	template<typename T>
-	struct offset_ptr_stl_allocator : std::allocator<T> {
+	template<typename T, template<typename> typename Allocator = std::allocator>
+	struct offset_ptr_stl_allocator {
 		using value_type = T;
 		using pointer = boost::interprocess::offset_ptr<T>;
+		using const_pointer = boost::interprocess::offset_ptr<T const>;
+		using void_pointer = boost::interprocess::offset_ptr<void>;
+		using const_void_pointer = boost::interprocess::offset_ptr<void const>;
 		using size_type = size_t;
 		using difference_type = std::ptrdiff_t;
+		using upstream_allocator_type = Allocator<T>;
 
-		using propagate_on_container_copy_assignment = typename std::allocator_traits<std::allocator<T>>::propagate_on_container_copy_assignment;
-		using propagate_on_container_move_assignment = typename std::allocator_traits<std::allocator<T>>::propagate_on_container_move_assignment;
-		using propagate_on_container_swap = typename std::allocator_traits<std::allocator<T>>::propagate_on_container_swap;
-		using is_always_equal = typename std::allocator_traits<std::allocator<T>>::is_always_equal;
+		using propagate_on_container_copy_assignment = typename std::allocator_traits<upstream_allocator_type>::propagate_on_container_copy_assignment;
+		using propagate_on_container_move_assignment = typename std::allocator_traits<upstream_allocator_type>::propagate_on_container_move_assignment;
+		using propagate_on_container_swap = typename std::allocator_traits<upstream_allocator_type>::propagate_on_container_swap;
+		using is_always_equal = typename std::allocator_traits<upstream_allocator_type>::is_always_equal;
 
-		constexpr offset_ptr_stl_allocator() noexcept = default;
+	private:
+		template<typename, template<typename> typename>
+		friend struct offset_ptr_stl_allocator;
+
+		[[no_unique_address]] upstream_allocator_type inner_;
+
+	public:
+		constexpr offset_ptr_stl_allocator() noexcept(std::is_nothrow_default_constructible_v<upstream_allocator_type>) = default;
```

Comment:
All constructions and assignments in this class should check if the underlying method is supported by upstream_allocator_type, e.g., here, if it is default constructible.